Update front coding text (#16491)
apache/druid: docs/ingestion/ingestion-spec.md

## `tuningConfig`

You specify tuning properties in a `tuningConfig`, which goes at the top level of an ingestion spec.

> Some properties apply to all [ingestion methods](./index.md#ingestion-methods), but most are specific to each individual ingestion method. An example `tuningConfig` that sets all of the shared, common properties to their defaults is:

Reviewer: I think it'll be helpful to explain the structure of an ingestion spec and what each component of the spec is. (Doesn't have to be this PR)

1. Type
2. Data schema
2a. Dimension spec
2b. Metrics spec
3. ioConfig
3a. inputSource
3b. inputFormat
4. tuningConfig

etc.
(Include things like transform specs etc. so that people can see the full structure is helpful)
Here we can explain to people that tuningConfig is about things that affects ingestion performance, hardware resource utilization, as well as changing storage behavior. etc.

### `indexSpec`

The `indexSpec` object can include the following properties.
For information on defining an `indexSpec` in a query context, see [SQL-based ingestion reference](../multi-stage-query/reference.md#context-parameters).

|Field|Description|Default|
|-----|-----------|-------|
|bitmap|Compression format for bitmap indexes. Should be a JSON object with `type` set to `roaring` or `concise`.|`{"type": "roaring"}`|
|dimensionCompression|Compression format for dimension columns. One of `lz4`, `lzf`, `zstd`, or `uncompressed`.|`lz4`|

> |stringDictionaryEncoding|Encoding format for STRING value dictionaries used by STRING and [COMPLEX&lt;json&gt;](../querying/nested-columns.md) columns. To enable front coding, set `stringDictionaryEncoding.type` to `frontCoded`. Optionally, you can specify the `bucketSize` and `formatVersion` properties. `bucketSize` is the number of values to place in a bucket to perform delta encoding. `bucketSize` defaults to 4. You can set it to any power of 2 less than or equal to 128. You use `formatVersion` to specify older versions for backwards compatibility during rolling upgrades. Options are `0` and `1`. `formatVersion` defaults to `0`. See [Front coding](#front-coding) for more information.|`{"type":"utf8"}`|

Author: Updated

|metricCompression|Compression format for primitive type metric columns. Options are `lz4`, `lzf`, `zstd`, `uncompressed`, or `none` (which is more efficient than `uncompressed`, but not supported by older versions of Druid).|`lz4`|
|longEncoding|Encoding format for long-typed columns. Applies regardless of whether they are dimensions or metrics. Options are `auto` or `longs`. `auto` encodes the values using offset or lookup table depending on column cardinality, and store them with variable size. `longs` stores the value as-is with 8 bytes each.|`longs`|
|jsonCompression|Compression format to use for nested column raw data. Options are `lz4`, `lzf`, `zstd`, or `uncompressed`.|`lz4`|

##### Front coding

:::info
Front coding is an [experimental feature](../development/experimental.md) introduced in Druid 25.0.0.
:::

Front coding is an incremental encoding strategy that lets you store STRING and [COMPLEX&lt;json&gt;](../querying/nested-columns.md) columns in Druid.

> With front coding enabled, Druid creates smaller UTF-8 encoded segments with very little performance cost.

Reviewer: ```suggestion
With front coding enabled, Druid creates smaller UTF-8 encoded segments with very little performance cost.
It's ideal for string columns where most of the front-part of the string is repeated, for example, URLs. URLs mostly starts with https://domain.xyz/directory/...., and front coding is able to exploit this pattern for more optimal compression when storing such datasets.
```

Author: Added

You can specify the following optional properties:
* `bucketSize`: Number of values to place in a bucket to perform delta encoding. Setting this property instructs indexing tasks to write segments using compressed dictionaries of the specified bucket size. You can set it to any power of 2 less than or equal to 128. `bucketSize` defaults to 4.
* `formatVersion`: Specifies older versions for backwards compatibility during rolling upgrades. Valid options are `0` and `1`. `formatVersion` defaults to `0`.

For example:

```
"indexSpec": {
  "stringDictionaryEncoding": {
    “type”:”frontCoded”,
    “bucketSize”: 4,
    “formatVersion”:0
  }
}
```

A new `"formatVersion": 1` was introduced in Druid 26.0.0, offering typically faster read speeds and smaller storage sizes.
When you upgrade from Druid 25.0.0 to Druid 26.0.0 while using front coding, the `formatVersion` defaults to `0` to keep writing out the older format. This default setting enables seamless downgrades to Druid 25.0.0 if needed.
It is recommended to enable the new format in a staging environment and thoroughly test your use case before deploying it to production.

:::caution

> By default, segments created with front coding enabled in Druid 26.0.0 are backwards compatible with Druid 25.0.0. Segments created with front coding in Druid 26.0.0 or 25.0.0 are not compatible with Druid versions older than 25.0.0.

Reviewer 2: ```suggestion
By default, segments created with front coding enabled in Druid 26.0.0 are backwards compatible with Druid 25.0.0. Segments created with front coding in Druid 26.0.0 or 25.0.0 are not compatible with Druid versions preceding 25.0.0.
```

A new `"formatVersion": 1` was introduced in Druid 26.0.0, offering typically faster read speeds and smaller storage sizes.
When you upgrade from Druid 25.0.0 to Druid 26.0.0 while using front coding, the `formatVersion` defaults to `0` to keep writing out the older format. This default setting enables seamless downgrades to Druid 25.0.0 if needed.
It is recommended to enable the new format in a staging environment and thoroughly test your use case before deploying it to production.

:::caution
By default, segments created with front coding enabled in Druid 26.0.0 are backwards compatible with Druid 25.0.0. Segments created with front coding in Druid 26.0.0 or 25.0.0 are not compatible with Druid versions older than 25.0.0.
:::

> Beyond these properties, each ingestion method has its own specific tuning properties. See the documentation for each

Reviewer 2: Did you remove this line intentionally? Could also be worth moving to intro, for example:

```
Some properties apply to all ingestion methods, but most are specific to a particular method.
See the documentation for each [ingestion method](./index.md#ingestion-methods) for details on the specific tuning properties for that method.

<newline>

The following example shows...
```

Author: I removed it because it seemed redundant. It links to the same source and the message is similar to 
`Some properties apply to all [ingestion methods](./index.md#ingestion-methods), but most are specific to each individual ingestion method.` Do you think I should add it back?

>     “formatVersion”:1

Author: Good catch!

|metricCompression|Compression format for primitive type metric columns. Options are `lz4`, `lzf`, `zstd`, `uncompressed`, or `none` (which is more efficient than `uncompressed`, but not supported by older versions of Druid).|`lz4`|
|longEncoding|Encoding format for long-typed columns. Applies regardless of whether they are dimensions or metrics. Options are `auto` or `longs`. `auto` encodes the values using offset or lookup table depending on column cardinality, and store them with variable size. `longs` stores the value as-is with 8 bytes each.|`longs`|
|jsonCompression|Compression format to use for nested column raw data. Options are `lz4`, `lzf`, `zstd`, or `uncompressed`.|`lz4`|

#### Front coding

:::info
Front coding is an [experimental feature](../development/experimental.md).
:::

Druid encodes string columns into dictionaries for better compression.
Front coding is an incremental encoding strategy that lets you store STRING and [COMPLEX&lt;json&gt;](../querying/nested-columns.md) columns in Druid with minimal performance impact.
Front-coded dictionaries reduce storage and improve performance by optimizing for strings where the front part looks similar.

> For example, if you are tracking website visits, most URLs start with `https://domain.xyz/`, and front coding is able to exploit this pattern for more optimal compression when storing such datasets.

Reviewer: ```suggestion
For example, if you are tracking website visits, most URLs start with `https://domain.xyz/`, and front coding is able to exploit this pattern for more optimal compression when storing such datasets. Durid does the optimization automatically, which means Druid usually doesn't negatively impact the performance of string columns when they don't match the front-coded pattern. Thus, allowing this feature to be turned on universally without you knowing the underlying column's data shapes.
```

Author: Updated.

> With front coding enabled, Druid tracks the length of common prefixes of values so that only the suffix is stored.

Reviewer: I don't think we need to talk about line 536/537 *(how it works internals)

Author: Removed.
